Update pull tests

icecrime opened the pull request:
This is a first step into improving code coverage of our push/pull code.
- Introduce the `DockerPushPullSuite` that interacts with its own dedicated daemon, thus allowing to start from a clean environment and finely test against the impact of isolated push and pull operations.
- Update and migrate existing tests to the `DockerPushPullSuite`. Most tests were preserved, but refactored and made more exhaustive. One test was deliberately removed (`TestPullVerified`) as it is unreliable and that the feature was obsoleted by content trust.
- Move all trust related tests to `docker_cli_pull_trusted_test.go`.
- Move tests depending on a local registry to `docker_cli_pull_local_test.go`.

After this PR gets in, I'll proceed in adding new test scenarios, and doing the same work on the push tests.

One thing I should mention (and feel free to file as an issue, or submit a PR) is that I'm not happy with the way `DockerPushPullSuite` is: it's too easy to incorrectly call to `dockerCmd` and hit the wrong daemon. I believe the global integration tests daemon started as part of `hack/make/.integration-daemon-start` should be modeled by a `Daemon` instance in the testing code, and that no function should execute commands without acting against a specific instance.

Commit:
Add test suite for push/pull code

Introduce the `DockerHubPullSuite` that interacts with its own dedicated
daemon, thus allowing to start from a clean environment and finely test
against the impact of isolated push and pull operations.

Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

Commit:
Update image pull tests

Update and migrate existing tests to the `DockerHubPullSuite`. Most
tests were preserved, but refactored and made more exhaustive. One test
was deliberately removed (`TestPullVerified`) as it is unreliable and
that the feature was obsoleted by content trust.

Move all trust related tests to `docker_cli_pull_trusted_test.go`.

Move tests depending on a local registry to `docker_cli_pull_local_test.go`.

Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

aaronlehmann commented:
> Introduce the DockerPushPullSuite that interacts with its own dedicated daemon, thus allowing to start from a clean environment and finely test against the impact of isolated push and pull operations.
Update and migrate existing tests to the DockerPushPullSuite. Most tests were preserved, but refactored and made more exhaustive. One test was deliberately removed (TestPullVerified) as it is unreliable and that the feature was obsoleted by content trust.

Awesome. I've run into issues with images not being cleaned up while working on push/pull tests. This will be a big improvement.

> One thing I should mention (and feel free to file as an issue, or submit a PR) is that I'm not happy with the way DockerPushPullSuite is: it's too easy to incorrectly call to dockerCmd and hit the wrong daemon. I believe the global integration tests daemon started as part of hack/make/.integration-daemon-start should be modeled by a Daemon instance in the testing code, and that no function should execute commands without acting against a specific instance.

Agreed.

aaronlehmann commented on integration-cli/push_pull_suite_test.go:
```diff
@@ -0,0 +1,83 @@
+type DockerPushPullSuite struct {
+	d  *Daemon
+	ds *DockerSuite
+}
```
As this is implemented now, the name `DockerPushPullSuite` is misleading, because this isn't usable for pushes (no private registry). I think we should do one of the following:

- This should be renamed to something like `DockerHubPullSuite`, and tests involving a private registry should be part of a different testsuite (possibly `DockerRegistrySuite`, with changes to make it similar to this testsuite).

or

- This testsuite could provide a private registry, and replace `DockerRegistrySuite`.

icecrime replied:
That's absolutely right, I'll rename to `DockerHubPullSuite`, thanks!

cpuguy83 commented:
ping @vdemeester Who is also working on cleaning up `dockerCmd` and other things.

vdemeester commented on integration-cli/docker_cli_pull_test.go:
```diff
@@ -2,126 +2,104 @@ package main
+	if matches := digestExpr.FindAllStringSubmatch(out, -1); len(matches) != 1 {
+		c.Fatalf("expected exactly one image digest in the output, got %d", len(matches))
+	} else if _, err := digest.ParseDigest(matches[0][1]); err != nil {
+		c.Fatalf("invalid digest %q in output", matches[0][1])
```
Probably just a nit (and because I think we don't use enough the power of `check`) — I'm writing this by memory so there might be a typo :sweat_smile: :

```go
matches = len(digestExpr.FindAllStringSubmatch(out, -1))
c.Check(matches, check.Equals, 1, check.Commentf("expected exactly one image digest in the output, got %d", matches))
_, err := digest.ParseDigest(matches[0][1])
c.Check(err, check.IsNil, check.Commentf("invalid digest %q in output", matches[0][1]))
```

LK4D4 replied:
I think you're right. @icecrime can try at least. Maybe it's not so easy as we think, but worth a try.

vdemeester commented on integration-cli/docker_cli_pull_test.go:
```diff
@@ -2,126 +2,104 @@ package main
+			c.Fatalf("expected non-zero exit status when pulling non-existing image: %s", out)
+		} else if !strings.Contains(out, fmt.Sprintf("Error: image %s not found", e.Image)) {
+			c.Fatalf("expected image not found error message: %s", out)
+		}
```
Same here for example :

```go
out, err := s.CmdWithError("pull", e.Alias)
c.Check(err, check.IsNil, check.Commentf("expected non-zero exit status when pulling non-existing image: %s", out))
c.Check(out, check.Matches, …, check.Commentf(…))
```

vdemeester commented:
Test looks good and I'm :+1: of adding more test suites that are smaller and make sense :wink:.

On the `dockerCmd` side these ones won't be a problem to fix as for almost all of them we don't use the return values (so no refactor will be needed on those).

> One thing I should mention (and feel free to file as an issue, or submit a PR) is that I'm not happy with the way DockerPushPullSuite is: it's too easy to incorrectly call to dockerCmd and hit the wrong daemon. I believe the global integration tests daemon started as part of hack/make/.integration-daemon-start should be modeled by a Daemon instance in the testing code, and that no function should execute commands without acting against a specific instance.

Agreed too.

icecrime commented:
I'm not happy with this yet, give me a moment ;-)

Commit:
Add integration-cli/checker package

Add a `checker` package that adds some utility Checker implementation,
the first one being `checker.Contains`, as well as brining all go-check
provided Checker implementations in scope as a commodity.

Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

icecrime commented:
Updated with much better use of `go-check`, as well as introducing our very own `integration-cli/checkers` package. PTAL!

aaronlehmann commented:
LGTM

vdemeester commented:
@icecrime this looks good :heart_eyes_cat:

jfrazelle commented:
LGTM

jfrazelle commented:
something is going on w windows its unrelated, LGTM
